packages/react-query/src/useBaseQuery.ts

```diff
@@ -82,33 +82,43 @@ export function useBaseQuery<
       ),
   )
 
+  const [_, setForceUpdate] = React.useState({})
+
   const result = observer.getOptimisticResult(defaultedOptions)
 
-  React.useSyncExternalStore(
-    React.useCallback(
-      (onStoreChange) => {
-        const unsubscribe = isRestoring
-          ? noop
-          : observer.subscribe(notifyManager.batchCalls(onStoreChange))
-
-        // Update result to make sure we did not miss any query updates
-        // between creating the observer and subscribing to it.
-        observer.updateResult()
-
-        return unsubscribe
-      },
-      [observer, isRestoring],
-    ),
-    () => observer.getCurrentResult(),
-    () => observer.getCurrentResult(),
-  )
+  React.useEffect(() => {
+    if (isRestoring) {
+      return
+    }
+
+    const unsubscribe = observer.subscribe(
+      notifyManager.batchCalls(() => {
+        setForceUpdate({})
+      }),
+    )
+
+    // Update result to make sure we did not miss any query updates
+    // between creating the observer and subscribing to it.
+    observer.updateResult()
+
+    return unsubscribe
+  }, [observer, isRestoring])
 
   React.useEffect(() => {
+    if (defaultedOptions.experimental_prefetchInRender) {
+      return
+    }
     // Do not notify on updates because of changes in the options because
     // these changes should already be reflected in the optimistic result.
     observer.setOptions(defaultedOptions, { listeners: false })
   }, [defaultedOptions, observer])
 
+  if (defaultedOptions.experimental_prefetchInRender) {
+    // Do not notify on updates because of changes in the options because
+    // these changes should already be reflected in the optimistic result.
+    observer.setOptions(defaultedOptions, { listeners: false })
```

Comment:
Setting the options straight away for `experimental_prefetchInRender` seems to fix it.

I think the issue I had was some race condition...

I still have no reliable reproduction though